### DISCUSSION TITLE
Workflow for exporting of animated 3D character from blender to Godot

### DISCUSSION THREAD

**Discussion by PerplexedPeach** [godotengine/godot-proposals#4680] [3D]
There are numerous issues with exporting an animated character from Blender to Godot (3.4.4.stable) and I'd like to document the problems, my current workarounds, and possible improvements. 

First off what I have in Blender (3.2):
- character rig with animations defined for it (includes non-deforming IK bones)
- multiple meshes (such as for body, hair, and clothes) with shape keys and drivers all parented to the same armature

What I need to export:
- meshes as separate objects (such as exporting body and hair separately) to allow easier mixing and matching of meshes
- skeleton to be exported with deforming bones only, as I found that non-deforming bones leads to visual issues with blend shapes
- animations of the rig

I first tried exporting the mesh and rig together through glTF with animations enabled. There are two major problems with this: 
1. drivers on the shape keys automatically creates NLA tracks that are not removable, and are exported along with the actual animation on the rig. 
![image](https://user-images.githubusercontent.com/48222967/174192494-8f6dce43-a051-49ef-8056-282f9fb43db6.png)
Removing the drivers presumably allows these tracks to be removed, however the drivers are needed for ease of posing. One possibility is duplicating the blend file, removing all the drivers, then exporting; however this needs to be done every time I need to export and is not viable.
3. the current value of shape keys are exported as part of the animation for each mesh 

I then tried deleting the NLA tracks prior to export, then undo to recover the animations. However, this did not work because the empty tracks from the drivers still caused an AnimationPlayer object to be instantiated on the exported .gltf. 
After this, I looked into options of exporting the meshes without animations at all; however, the option to export only deforming bones is only available if the animation is checked in the export menu:
![image](https://user-images.githubusercontent.com/48222967/174193104-e40e593b-bcb9-4e66-9e7a-09c3ca9b94b0.png)

I now export the meshes independently with non-deforming bones, then export just the rig with animations and only deforming bones. Importing both of these into Godot, I then transfer the MeshInstances from the Skeleton(s) with non-deforming bones over to the Skeleton with only deforming bones, as well as adding that Skeleton's parent node to the scene tree:
```swift
# the mesh skeletons have non-deforming bones, so we need to transfer the meshes
# to the animation skeleton, which was exported with only deforming bones
onready var sk_meshes:Skeleton = $"Daemon Girl_rig/Skeleton"
onready var sk_anim_scene = preload("res://player/anim/player_anim.tscn").instance()
onready var sk:Skeleton = sk_anim_scene.get_node("Daemon Girl_rig/Skeleton");
onready var sk_anim:AnimationPlayer = sk_anim_scene.get_node("AnimationPlayer");

func loadBodyParts():
	get_parent().call_deferred("add_child", sk_anim_scene)
	body=sk_meshes.get_node_or_null("body")
	eyeLash = sk_meshes.get_node_or_null("eyelash")
	eye = sk_meshes.get_node_or_null("eyes")
	eyeCornea = sk_meshes.get_node_or_null("cornea")
	# transfer them to the clean skeleton from the rig
	partsHead = [body, eyeLash, eye, eyeCornea]
	for part in partsHead:
		sk_meshes.remove_child(part)
		sk.add_child(part)
	sk_meshes.queue_free()
```

Where this code is attached to the `player` scene, which is directly linked from the active game scene
![image](https://user-images.githubusercontent.com/48222967/174193661-dd6a165d-b676-40ff-a767-a8f7bd5bad56.png)

While thing seem to work so far, the loading is very slow and I am getting errors because I am transferring skin binds that have non-deforming bones to a skeleton that has only deforming bones, leading to hundreds of errors:
![image](https://user-images.githubusercontent.com/48222967/174193778-1a693636-551d-4441-85d4-93d6611de4ca.png)

How can I improve this situation, such as by improving the load speed, making the export process easier, or removing the errors? Is it much easier to just do all the animations in Godot? Even after the grueling 10+ hours experimenting with exporting the animation, it is still incorrect (forearm is too thin for whatever reason):
![image](https://user-images.githubusercontent.com/48222967/174196765-acb44907-b1bb-4db6-8267-38db31d0dd8d.png)
![image](https://user-images.githubusercontent.com/48222967/174196810-e2a59cdc-7750-4840-a820-ec05ffc6e76b.png)

**Comment by PerplexedPeach** [comment]
The incorrect forearm stretching seems to be because in Blender, no keyframes were assigned to the `forearm_stretch.l` bone 
![image](https://user-images.githubusercontent.com/48222967/174390347-c534fbad-9ff0-42db-b17c-a8e459a3a084.png)
But it does exist in the animation exported to Godot:
![image](https://user-images.githubusercontent.com/48222967/174390564-5eb6ffbe-4d4b-4579-ab79-58979ba61691.png)

I'll try redefining the animation to include these bones. What could be happening is that inside Blender, the key frames are mainly defined on the IK control bones, so even though the animations look fine, they are broken after exporting since only deforming bones matter after export.

**Reply by lentsius-bark** [reply]
Keyframing bones to fix things on Godot import has worked for me before in the past as well (ok in Godot). The experience of importing from blender seems solid for simple stuff but when bones are involved I am never 100% sure it'll work as expected/won't break on reimport.

**Comment by PerplexedPeach** [comment]
Yeah, unfortunately keyframing the deforming bones did not fix the issue. Pausing the animations inside Godot still results in the thin forearm (but that doesn't mean that much since it holds the value of the bones last played I believe).

Something else I tried that did not improve things is cleaning up vertex groups (removing low influence vertex groups for each vertex). I believe glTF export only exports 4 vertex groups at most per vertex (not sure how it selects which ones to export, but I suspect it chooses the highest influencing ones). Choosing to export with all bone influence under skinning will result in the mesh not showing up in Godot by the way:
![image](https://user-images.githubusercontent.com/48222967/174395056-d7938879-fb59-4b36-acee-a3657ebcb4e3.png)

Also a note is that if you have non-deforming vertex groups such as Smoothing, or Physics, you should normalize all weights with the deforming subset only, as otherwise it will include those groups that don't correspond to deforming bones.

Good news is that there is a new pull request for the glTF exporter for separating the export of deforming only bones away from exporting animations: https://github.com/KhronosGroup/glTF-Blender-IO/issues/1668

**Reply by PerplexedPeach** [reply]
The pull request was merged so should be in Blender 3.3
